pyembed/src/importer.rs

```diff
@@ -282,99 +276,89 @@ pub(crate) struct ImporterState {
-                py,
+        let meta_path = meta_path_object.cast_as::<PyList>()?;
+        if meta_path.len() < 2 {
+            return Err(PyValueError::new_err(
                 "sys.meta_path does not contain 2 values",
             ));
         }
 
-        let builtin_importer = meta_path.get_item(py, 0);
-        let frozen_importer = meta_path.get_item(py, 1);
+        let builtin_importer = meta_path.get_item(0).into_py(py);
+        let frozen_importer = meta_path.get_item(1).into_py(py);
 
-        let marshal_loads = marshal_module.get(py, "loads")?;
-        let call_with_frames_removed = bootstrap_module.get(py, "_call_with_frames_removed")?;
-        let module_spec_type = bootstrap_module.get(py, "ModuleSpec")?;
+        let marshal_loads = marshal_module.getattr("loads")?.into_py(py);
+        let call_with_frames_removed = bootstrap_module
+            .getattr("_call_with_frames_removed")?
+            .into_py(py);
+        let module_spec_type = bootstrap_module.getattr("ModuleSpec")?.into_py(py);
 
-        let builtins_module = match unsafe {
-            cpython::PyObject::from_borrowed_ptr_opt(py, oldpyffi::PyEval_GetBuiltins())
-        } {
-            Some(o) => o.cast_into::<cpython::PyDict>(py),
-            None => {
-                return Err(cpython::PyErr::new::<cpython::exc::ValueError, _>(
-                    py,
-                    "unable to obtain __builtins__",
-                ));
-            }
-        }?;
+        let builtins_module =
+            unsafe { PyDict::from_borrowed_ptr_or_err(py, pyffi::PyEval_GetBuiltins()) }?;
 
-        let exec_fn = match builtins_module.get_item(py, "exec") {
+        let exec_fn = match builtins_module.get_item("exec") {
             Some(v) => v,
             None => {
-                return Err(cpython::PyErr::new::<cpython::exc::ValueError, _>(
-                    py,
-                    "could not obtain __builtins__.exec",
-                ));
+                return Err(PyValueError::new_err("could not obtain __builtins__.exec"));
             }
-        };
+        }
+        .into_py(py);
 
-        let sys_flags = sys_module.get(py, "flags")?;
-        let optimize_value = sys_flags.getattr(py, "optimize")?;
-        let optimize_value = optimize_value.extract::<i64>(py)?;
+        let sys_flags = sys_module.getattr("flags")?;
+        let sys_module = sys_module.into_py(py);
+
+        let optimize_value = sys_flags.getattr("optimize")?;
+        let optimize_value = optimize_value.extract::<i64>()?;
 
         let optimize_level = match optimize_value {
             0 => Ok(OptimizeLevel::Zero),
             1 => Ok(OptimizeLevel::One),
             2 => Ok(OptimizeLevel::Two),
-            _ => Err(cpython::PyErr::new::<cpython::exc::ValueError, _>(
-                py,
+            _ => Err(PyValueError::new_err(
                 "unexpected value for sys.flags.optimize",
             )),
         }?;
 
         let capsule = unsafe {
```

Comment:
I'd be interested in potentially working through https://github.com/PyO3/pyo3/issues/1193 sometime, looks like you might want it (and / or want to weigh in on its design).